src/main/cpp/src/number_converter.cu

```diff
@@ -0,0 +1,560 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "number_converter.hpp"
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/indexalator.cuh>
+#include <cudf/detail/sizes_to_offsets_iterator.cuh>
+#include <cudf/detail/utilities/assert.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/types.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/logical.h>
+#include <thrust/pair.h>
+
+#include <cstdlib>  // For abs() function
+
+namespace spark_rapids_jni {
+
+namespace {
+
+constexpr int MIN_BASE = 2;
+constexpr int MAX_BASE = 36;
+
+/**
+ * @brief Trims space characters (ASCII 32)
+ * @return The first non-space index and last non-space index pair
+ */
+__device__ thrust::pair<int, int> trim(char const* ptr, int len)
+{
+  int first = 0;
+  int last  = len - 1;
+  while (first < len && ptr[first] == ' ') {
+    ++first;
+  }
+  while (last > first && ptr[last] == ' ') {
+    --last;
+  }
+  return thrust::make_pair(first, last);
+}
+
+/**
+ * @brief Convert byte value to char, assume base is in range [0, 35]
+ * E.g.:
+ *   0 => '0'
+ *   ...
+ *   9 => '9'
+ *   10 => 'A'
+ *   ...
+ *   15 => 'F'
+ *   ...
+ *   35 => 'Z'
+ * Note:
+ *   For base ragne in [2, 36], the max char is 'Z'
+ */
+__device__ char byte_to_char(int byte_value)
+{
+  if (byte_value >= 0 && byte_value <= 9) {
+    return static_cast<char>('0' + (byte_value - 0));
+  } else if (byte_value >= 10 && byte_value <= 35) {
+    return static_cast<char>('A' + (byte_value - 10));
+  } else {
+    cudf_assert(false);
+  }
+}
+
+/**
+ * @brief Convert char to byte value, it's a reverse of byte_to_char; If char is invalid
+ * considering 'base', return -1.
+ * E.g.: base = 36
+ *   '0' => 0
+ *   ...
+ *   '9' => 9
+ *   'A' or 'a' => 10
+ *   ...
+ *   'F' or 'f' => 15
+ *   ...
+ *   'Z' or 'z' => 35
+ * E.g.: base = 8
+ *   '8' => -1 // invalid char considering base
+ */
+__device__ int char_to_byte(char c, int base)
+{
+  if (c >= '0' && c <= '9' && c - '0' < base) {
+    return c - '0';
+  } else if (c >= 'A' && c <= 'Z' && c - 'A' + 10 < base) {
+    return c - 'A' + 10;
+  } else if (c >= 'a' && c <= 'z' && c - 'a' + 10 < base) {
+    return c - 'a' + 10;
+  } else {
+    return -1;
+  }
+}
```

Comment:
Should  we compute 

char-to-byte 

and 
byte-to-char

as 256-element arrays once and lookup mapped values without branching?

Maybe even as constexpr as specialization for common bases 2, 10 and 16 ?